### ISSUE TITLE
Why models are saved with half precision?

### ISSUE THREAD

**Issue by Hyunseok-Kim0** [ultralytics/ultralytics#3089]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

https://github.com/ultralytics/ultralytics/blob/3ae81ee9d11c432189e109d7a1724635a2e451ca/ultralytics/yolo/engine/trainer.py#L405-L416

`save_model` uses .half() when saving models. This is usually not a matter but still the performance degradation definitely exists. Half precision reduced about 0.03 - 0.07 mAP in my cases. 
What I think is a real problem is this is default option which is not exposed. There can be a lot of people who try to find out the reason in different part that performance of training log is not matched with validation result. This also makes problem when pause and restart the training. The user will experience unexpected mAP drop even though there was no special thing has done. Is there any special reason saving models with half precision?

**Comment by Hyunseok-Kim0** [comment]
https://colab.research.google.com/gist/Hyunseok-Kim0/f25fbda51bd3232daf07224fa1021ec6/yolov8-half-test.ipynb

Here is sample gist. I trained and validated yolov8n model for coco128 just to figure out about performance drop. mAP50-95 changed from 0.44944831379869 to 0.4493777580801172 when .half() is used. The difference (0.000070556) is ㅡmuch smaller than I experienced before. Maybe I mistook the numbers.

**Comment by glenn-jocher** [comment]
Hello @Hyunseok-Kim0,

Thank you for sharing your findings with us. The reason why models are saved with half precision is to reduce the memory footprint and enable larger model architectures or batch sizes to be used during training. Additionally, it should not cause any detrimental effect to the accuracy of the model. 

However, we understand that you experienced a decrease in mAP50-95 values when using .half(). Our recommendation is to re-run the experiments and ensure that all parameters between the two runs are consistent. Inconsistencies such as learning rate, optimizer, data augmentation, and random seeds can all contribute to variations in model performance.

Please keep us updated and let us know if you have any further concerns.

Best, 
Glenn Jocher

**Comment by Hyunseok-Kim0** [comment]
https://github.com/ultralytics/ultralytics/blob/f4e8b39fc2ac4528c8378ec077e8df4afe329f60/ultralytics/yolo/engine/trainer.py#L439-L454
https://github.com/ultralytics/ultralytics/blob/f4e8b39fc2ac4528c8378ec077e8df4afe329f60/ultralytics/nn/tasks.py#L529-L536

>Thank you for sharing your findings with us. The reason why models are saved with half precision is to reduce the memory footprint and enable larger model architectures or batch sizes to be used during training. Additionally, it should not cause any detrimental effect to the accuracy of the model.

As seen above `attempt_load_weights`, the model use .float() when loading weights in trainer. It looks the model is trained on single precision and saved to half precision and then loaded to single precision again. The model always run on float32, is it really helpful to reducing the memory usage saving the model using half precision? 

Also, as you can see in the gist I linked, I did not change any parameter between two validation. The model shows different validation results only after calling .half().

**Comment by pderrenger** [comment]
@Hyunseok-Kim0 thank you for the detailed insight into the specific areas of the codebase. The usage of `.half()` and `.float()` is intended to reduce memory usage when the model is saved while using half precision, and then loaded and run on float32 precision. The performance degradation you observed with `.half()` is unexpected.

We'll investigate this issue further to understand the root cause and address any unexpected behavior related to model precision. Please feel free to open an issue in the repository if one does not currently exist to track this specific behavior, or contribute further to the ongoing discussions.
